include/proxy-wasm/context.h

```diff
@@ -0,0 +1,736 @@
+   */
+  virtual WasmResult grpcStream(string_view /* grpc_service */, string_view /* service_name */,
+                                string_view /* method_name */, uint32_t * /* token_ptr */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Close a gRPC stream.  In flight data may still result in calls into the VM.
+   * @param token is a token returned from grpcStream.
+   */
+  virtual WasmResult grpcClose(uint32_t /* token */) { // cancel on call, close on stream.
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Cancel a gRPC stream or call.  No more calls will occur.
+   * @param token is a token returned from grpcSream or grpcCall.
+   */
+  virtual WasmResult grpcCancel(uint32_t /* token */) { // cancel on call, reset on stream.
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Close a gRPC stream.  In flight data may still result in calls into the VM.
+   * @param token is a token returned from grpcSream.
+   * @param message is a (serialized) message to be sent.
+   * @param end_stream indicates that the stream is now end_of_stream (e.g. WriteLast() or
+   * WritesDone).
+   */
+  virtual WasmResult grpcSend(uint32_t /* token */, string_view /* message */,
+                              bool /* end_stream */) { // stream only
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  // Metrics
+
+  /**
+   * Define a metric (Stat).
+   * @param type is the type of metric (e.g. Counter).
+   * @param name is a string uniquely identifying the metric.
+   * @param metric_id_ptr is a location to store a token used for subsequent operations on the
+   * metric.
+   */
+  virtual WasmResult defineMetric(MetricType /* type */, string_view /* name */,
+                                  uint32_t * /* metric_id_ptr */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Increment a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param offset is the offset to apply to the Counter.
+   */
+  virtual WasmResult incrementMetric(uint32_t /* metric_id */, int64_t /* offset */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Record a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param value is the value to store for a Gauge or increment for a histogram or Counter.
+   */
+  virtual WasmResult recordMetric(uint32_t /* metric_id */, uint64_t /* value */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Get the current value of a metric.
+   * @param metric_id is a token returned by defineMetric() identifying the metric to operate on.
+   * @param value_ptr is a location to store the value of the metric.
+   */
+  virtual WasmResult getMetric(uint32_t /* metric_id */, uint64_t * /* value_ptr */) {
```

Comment:
What does this mean for histograms? Per the conversation in the Envoy repo, you should be able to get histogram values and will certainly need to do this for stat plugins in the future. This is related to my comment above about splitting out the metric types.

Reply:
This is what Joshua Marantz said about getting histogram values:

**TBH I'm not sure what you'll get with those. The reason is that the histogram data is accumulated in thread-local caches, and it only gets aggregated in the 'parent histogram' in response to some event; idk if that's interval based or just when the histogram data is polled from the admin interface.

It should be possible to add some kind of API to initiate a cross-thread aggregation and then call a callback with the aggregated values. But at any given point in time there will not be an all-thread aggregation available. This interface would obviously be fairly expensive to implement.**

Reply:
Yeah I have to admit I haven't used the histograms much.

Now that I see the context, think @mattklein123 meant that a getMetric() API that synchronously returns a single scalar value for a histogram doesn't make that much sense.

We should probably have an API that returns a bucket-array in some form, and then there's the question of when that data is aggregated from the threads.

@jplevyak what data would you want out of the histograms?

Reply:
Yes this is my point. You can get histograms out of Envoy's stat store today, this is how we emit histograms to stat sinks. My point still stands, even setting Envoy aside, there has to be some understanding of how counters/gauges/histogram like things are modeled. Perhaps that ABI should model counters/gauges only for now and histogram, etc. is proxy specific? Or it should be documented like I asked that certain combinations of parameters will never work because they don't make sense? Etc.

Reply:
The combination of parameters which do not work is something which we need to document in the SDK documentation.   I don't think it is a function of this layer since it depends on the specific implementation details of the stats system which may change over time.   We have changed the way that the envoy-wasm code uses stats several times already, so I don't doubt we will do so again, and the specifics of those limitations are more interesting to end-users.   At this level suffice that recordMetric will do the logical thing for all types of metrics.

Reply:
The problem is `recordMetric` might not do the right thing for all metrics. Let's chat about this in the spec PR and/or our meeting tomorrow.

Reply:
Discussed offline.  This will be covered int the spec.